hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java

```diff
@@ -33,88 +33,82 @@
- *
- * NOTE : Values are only appended to disk. If a remove() is called, the entry is marked removed from the in-memory
- * key-valueMetadata map but it's values will be lying around in the temp file on disk until the file is cleaned.
- *
- * The setting of the spill threshold faces the following trade-off: If the spill threshold is
- * too high, the in-memory map may occupy more memory than is available, resulting in OOM.
- * However, if the spill threshold is too low, we spill frequently and incur unnecessary disk
- * writes.
- * @param <T>
- * @param <R>
+ * An external map that spills content to disk when there is insufficient space for it to grow. <p>
+ * This map holds 2 types of data structures : <p> (1) Key-Value pairs in a in-memory map (2)
+ * Key-ValueMetadata pairs in an in-memory map which keeps a marker to the values spilled to disk
+ * <p> NOTE : Values are only appended to disk. If a remove() is called, the entry is marked removed
+ * from the in-memory key-valueMetadata map but it's values will be lying around in the temp file on
+ * disk until the file is cleaned. <p> The setting of the spill threshold faces the following
+ * trade-off: If the spill threshold is too high, the in-memory map may occupy more memory than is
+ * available, resulting in OOM. However, if the spill threshold is too low, we spill frequently and
+ * incur unnecessary disk writes.
  */
-public class ExternalSpillableMap<T,R> implements Map<T,R> {
+public class ExternalSpillableMap<T, R> implements Map<T, R> {
 
   // maximum space allowed in-memory for this map
   final private long maxInMemorySizeInBytes;
   // current space occupied by this map in-memory
   private Long currentInMemoryMapSize;
   // Map to store key-values in memory until it hits maxInMemorySizeInBytes
-  final private Map<T,R> inMemoryMap;
+  final private Map<T, R> inMemoryMap;
   // Map to store key-valuemetadata important to find the values spilled to disk
-  final private DiskBasedMap<T,R> diskBasedMap;
-  // Schema used to de-serialize and readFromDisk the records written to disk
-  final private Schema schema;
+  final private DiskBasedMap<T, R> diskBasedMap;
   // An estimate of the size of each payload written to this map
   private volatile long estimatedPayloadSize = 0;
   // TODO(na) : a dynamic sizing factor to ensure we have space for other objects in memory and incorrect payload estimation
   final private Double sizingFactorForInMemoryMap = 0.8;
+  // Key converter to convert key type to bytes
+  final private Converter<T> keyConverter;
+  // Value converter to convert value type to bytes
+  final private Converter<R> valueConverter;
+  // Find the actual estimated payload size after inserting N records
+  final private static int NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE = 100;
+  // Flag to determine whether to stop re-estimating payload size
+  private boolean shouldEstimatePayloadSize = true;
```

Comment:
Should we eliminate this flag and just reestimate the hashmap value every X records or so continuously? It can be non-linear probe as well.. 100, 1000, 10000, 100000 to amortize cost?

Reply:
Yeah, I thought of that but I'm unsure of the performance of the ObjectSizeEstimator for such large objects, say a HashMap with 100K entries. Do you think we should performance test that now and do this continuously ?